### ISSUE TITLE
vector panicked when reading from kafka

### ISSUE THREAD

**Issue by jeromekleinen-kbc** [vectordotdev/vector#10737]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Vector Version

```
vector 0.19.0 (x86_64-unknown-linux-gnu da60b55 2021-12-28) (same behavior on 0.18 though)
```

### Vector Configuration File

```
kafka source with decoding.codec = "json"
console sink
```

### Debug Output

```
[user@server vector]$ RUST_BACKTRACE=1 RUST_BACKTRACE=full vector -vvv -c aeh_dvcfe.toml
2022-01-06T13:07:19.223929Z  INFO vector::app: Log level is enabled. level="vector=trace,codec=trace,vrl=trace,file_source=trace,tower_limit=trace,rdkafka=trace,buffers=trace"
2022-01-06T13:07:19.224184Z  INFO vector::app: Loading configs. paths=["aeh_dvcfe.toml"]
2022-01-06T13:07:19.233902Z TRACE log: Create new librdkafka client 0x55c6f6edb0d0 log.target="rdkafka::client" log.module_path="rdkafka::client" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/client.rs" log.line=201
2022-01-06T13:07:19.234011Z TRACE log: Destroying topic partition list: 0x55c6f6fc8230 log.target="rdkafka::util" log.module_path="rdkafka::util" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/util.rs" log.line=264
2022-01-06T13:07:19.234032Z TRACE log: Destroyed topic partition list: 0x55c6f6fc8230 log.target="rdkafka::util" log.module_path="rdkafka::util" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/util.rs" log.line=266
2022-01-06T13:07:19.234121Z TRACE log: Starting stream consumer wake loop: 0x55c6f6edb0d0 log.target="rdkafka::consumer::stream_consumer" log.module_path="rdkafka::consumer::stream_consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/stream_consumer.rs" log.line=284
2022-01-06T13:07:19.234196Z  INFO vector::topology::running: Running healthchecks.
2022-01-06T13:07:19.234241Z  INFO vector::topology::running: Starting source. key=aeh
2022-01-06T13:07:19.234276Z  INFO vector::topology::builder: Healthcheck: Passed.
2022-01-06T13:07:19.234296Z  INFO vector::topology::running: Starting sink. key=my_sink_id
2022-01-06T13:07:19.234341Z  INFO vector: Vector has started. debug="false" version="0.19.0" arch="x86_64" build_id="da60b55 2021-12-28"
2022-01-06T13:07:19.234362Z  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
2022-01-06T13:07:19.235292Z TRACE vector: Beep.
2022-01-06T13:07:19.235299Z DEBUG sink{component_kind="sink" component_id=my_sink_id component_type=console component_name=my_sink_id}: vector::utilization: utilization=0.01875685565284768
2022-01-06T13:07:20.235372Z TRACE vector: Beep.
2022-01-06T13:07:21.235164Z TRACE vector: Beep.
2022-01-06T13:07:22.235325Z TRACE vector: Beep.
2022-01-06T13:07:23.235324Z TRACE vector: Beep.
2022-01-06T13:07:24.039142Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Running pre-rebalance with Assign(TPL {topicname/0: offset=Invalid metadata=""; topicname/1: offset=Invalid metadata=""; topicname/2: offset=Invalid metadata=""; topicname/3: offset=Invalid metadata=""}) log.target="rdkafka::consumer" log.module_path="rdkafka::consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/mod.rs" log.line=69
2022-01-06T13:07:24.039198Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Running rebalance with Assign(TPL {topicname/0: offset=Invalid metadata=""; topicname/1: offset=Invalid metadata=""; topicname/2: offset=Invalid metadata=""; topicname/3: offset=Invalid metadata=""}) log.target="rdkafka::consumer" log.module_path="rdkafka::consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/mod.rs" log.line=72
2022-01-06T13:07:24.039348Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Running post-rebalance with Assign(TPL {topicname/0: offset=Invalid metadata=""; topicname/1: offset=Invalid metadata=""; topicname/2: offset=Invalid metadata=""; topicname/3: offset=Invalid metadata=""}) log.target="rdkafka::consumer" log.module_path="rdkafka::consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/mod.rs" log.line=96
2022-01-06T13:07:24.235470Z TRACE vector: Beep.
2022-01-06T13:07:24.235485Z DEBUG sink{component_kind="sink" component_id=my_sink_id component_type=console component_name=my_sink_id}: vector::utilization: utilization=0.0018757667423237585
2022-01-06T13:07:25.235672Z TRACE vector: Beep.
2022-01-06T13:07:25.342176Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: vector::internal_events::kafka: Received one event. internal_log_rate_secs=10
thread 'vector-worker' panicked at 'event must exist', src/sources/kafka.rs:238:58
stack backtrace:
   0:     0x55c6f51a475c - std::backtrace_rs::backtrace::libunwind::trace::hf6a6dfd7da937cb0
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x55c6f51a475c - std::backtrace_rs::backtrace::trace_unsynchronized::hc596a19e4891f7f3
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55c6f51a475c - std::sys_common::backtrace::_print_fmt::hb16700db31584325
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x55c6f51a475c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h231c4190cfa75162
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x55c6f52009dc - core::fmt::write::h2a1462b5f8eea807
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/fmt/mod.rs:1163:17
   5:     0x55c6f51948c5 - std::io::Write::write_fmt::h71ddfebc68685972
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/io/mod.rs:1696:15
   6:     0x55c6f51a7c20 - std::sys_common::backtrace::_print::hcc197d4bebf2b369
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x55c6f51a7c20 - std::sys_common::backtrace::print::h335a66af06738c7c
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x55c6f51a7c20 - std::panicking::default_hook::{{closure}}::h6fac9ac9c8b79e52
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:210:50
   9:     0x55c6f51a77d5 - std::panicking::default_hook::h341c1030c6a1161b
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:227:9
  10:     0x55c6f51a8414 - std::panicking::rust_panic_with_hook::h50680ff4b44510c6
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:624:17
  11:     0x55c6f51a7ef0 - std::panicking::begin_panic_handler::{{closure}}::h9371c0fbb1e8465a
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:521:13
  12:     0x55c6f51a4c04 - std::sys_common::backtrace::__rust_end_short_backtrace::h9b3efa22a5768c0f
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:139:18
  13:     0x55c6f51a7e59 - rust_begin_unwind
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
  14:     0x55c6f0c15ed1 - core::panicking::panic_fmt::h23b9203e89cc61cf
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
  15:     0x55c6f51fd871 - core::panicking::panic_display::h0b2b4cc4e8e79e2e
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:64:5
  16:     0x55c6f0c15dcb - core::option::expect_failed::h111c62e88059319b
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/option.rs:1638:5
  17:     0x55c6f0e783e0 - <futures_util::stream::stream::filter_map::FilterMap<St,Fut,F> as futures_core::stream::Stream>::poll_next::h574e8bbecade1a6c
  18:     0x55c6f22735af - vector::sources::kafka::kafka_source::{{closure}}::h4bfe327d204e828f
  19:     0x55c6f1d472b6 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h44766d1dd7d26a8e
  20:     0x55c6f1e653f0 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h9c1ddec333a5bb3c
  21:     0x55c6f149fde6 - tokio::runtime::task::harness::poll_future::h5bbae3724bbf634a
  22:     0x55c6f1442885 - tokio::runtime::task::raw::poll::h29806d2e517b3f8f
  23:     0x55c6f4edd474 - tokio::runtime::thread_pool::worker::Context::run_task::hcc87b824b4a63258
  24:     0x55c6f4edc827 - tokio::runtime::thread_pool::worker::run::hed6d47bd2b49ef6f
  25:     0x55c6f4eccce4 - tokio::runtime::task::raw::poll::h4edb309ccd503a22
  26:     0x55c6f4e9f646 - std::sys_common::backtrace::__rust_begin_short_backtrace::h97657e221c44307c
  27:     0x55c6f4ea6c74 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h802ad6f352b8a433
  28:     0x55c6f51b35f3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h7bd677a5dc988be6
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/alloc/src/boxed.rs:1691:9
  29:     0x55c6f51b35f3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h7b1c1ba11c4db785
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/alloc/src/boxed.rs:1691:9
  30:     0x55c6f51b35f3 - std::sys::unix::thread::Thread::new::thread_start::h9c58c0d12d84e854
                               at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys/unix/thread.rs:106:17
  31:     0x7f771f7a914a - start_thread
  32:     0x7f771e586dc3 - clone
  33:                0x0 - <unknown>
2022-01-06T13:07:25.367271Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Destroying message: 0x7f76f8015488 log.target="rdkafka::util" log.module_path="rdkafka::util" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/util.rs" log.line=264
2022-01-06T13:07:25.367299Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Destroyed message: 0x7f76f8015488 log.target="rdkafka::util" log.module_path="rdkafka::util" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/util.rs" log.line=266
2022-01-06T13:07:25.367338Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Destroying consumer: 0x55c6f6edb0d0 log.target="rdkafka::consumer::base_consumer" log.module_path="rdkafka::consumer::base_consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/base_consumer.rs" log.line=581
2022-01-06T13:07:25.367365Z  INFO vector::shutdown: All sources have finished.
2022-01-06T13:07:25.367380Z  INFO vector: Vector has stopped.
2022-01-06T13:07:25.367685Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="aeh, my_sink_id" time_remaining="59 seconds left"
2022-01-06T13:07:25.370367Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Running pre-rebalance with Revoke log.target="rdkafka::consumer" log.module_path="rdkafka::consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/mod.rs" log.line=69
2022-01-06T13:07:25.370396Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Running rebalance with Revoke log.target="rdkafka::consumer" log.module_path="rdkafka::consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/mod.rs" log.line=72
2022-01-06T13:07:25.370841Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Running post-rebalance with Revoke log.target="rdkafka::consumer" log.module_path="rdkafka::consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/mod.rs" log.line=96
2022-01-06T13:07:25.388142Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Consumer destroyed: 0x55c6f6edb0d0 log.target="rdkafka::consumer::base_consumer" log.module_path="rdkafka::consumer::base_consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/base_consumer.rs" log.line=583
2022-01-06T13:07:25.388175Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Destroying client: 0x55c6f6edb0d0 log.target="rdkafka::util" log.module_path="rdkafka::util" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/util.rs" log.line=264
2022-01-06T13:07:25.389285Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Destroyed client: 0x55c6f6edb0d0 log.target="rdkafka::util" log.module_path="rdkafka::util" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/util.rs" log.line=266
2022-01-06T13:07:25.389314Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Destroying queue: 0x55c6f6fc8210 log.target="rdkafka::util" log.module_path="rdkafka::util" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/util.rs" log.line=264
2022-01-06T13:07:25.389329Z TRACE source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: log: Destroyed queue: 0x55c6f6fc8210 log.target="rdkafka::util" log.module_path="rdkafka::util" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/util.rs" log.line=266
2022-01-06T13:07:25.389387Z ERROR source{component_kind="source" component_id=aeh component_type=kafka component_name=aeh}: vector::topology: An error occurred that vector couldn't handle.
2022-01-06T13:07:25.389385Z TRACE log: Shut down stream consumer wake loop: 0x55c6f6edb0d0 log.target="rdkafka::consumer::stream_consumer" log.module_path="rdkafka::consumer::stream_consumer" log.file="/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/consumer/stream_consumer.rs" log.line=291
2022-01-06T13:07:25.389462Z DEBUG sink{component_kind="sink" component_id=my_sink_id component_type=console component_name=my_sink_id}: vector::topology::builder: Finished.
```

### Expected Behavior

No panic

### Actual Behavior

Panic

### Additional Context

Not sure if relevant, but the source is an azure event hub, but this should be kafka compatible. We use vector to consume multiple azure event hubs without any issues. This specific event hub contains batched events (an array of events in a single message). The json decoder seems to be able to handle this by emitting multiple events (tested using the stdin source), similar to the VRL behavior, but this might be the cause of the issue.

**Comment by jeromekleinen-kbc** [comment]
As a workaround, it seems that removing the decoding.codec setting and passing the data to a simple remap transform which calls parse_json does manage to parse the message into different events. The size could maybe also be an issue as some of these messages are close to the 1mb message size limit of azure event hub.

**Comment by spencergilbert** [comment]
@fuchsnj, `git blame` points towards you, do you know how we could run into this?

**Comment by fuchsnj** [comment]
@spencergilbert yes, I'll submit a fix shortly.

**Comment by jeromekleinen-kbc** [comment]
@jszwedko Would it be possible to add this fix to the 0.19.1 milestone? It has already landed on master.

**Comment by jszwedko** [comment]
Hi @jeromekleinen-kbc !

I did take a look at this issue to see if it should be in 0.19.1. Typically we'd only include regressions in the last minor release  in subsequent patch releases to reduce risk, and this issue existed in 0.18. I'll take another look, but we might hold onto this one until 0.20.0 which would be on Feb 8th.
